crates/p2p/src/client/peer_agnostic.rs

```diff
@@ -187,6 +193,186 @@ impl Client {
             .send_transactions_sync_request(peer, request)
             .await
     }
+
+    /// ### Important
+    ///
+    /// Caller must guarantee that `start <= stop_inclusive`
+    pub fn contract_updates_stream(
+        self,
+        mut start: BlockNumber,
+        stop_inclusive: BlockNumber,
+        getter: Arc<
+            impl Fn(
+                    BlockNumber,
+                    NonZeroUsize,
+                ) -> anyhow::Result<Option<SmallVec<[StateUpdateStats; 10]>>>
+                + Send
+                + Sync
+                + 'static,
+        >,
+    ) -> impl futures::Stream<Item = anyhow::Result<PeerData<(BlockNumber, ContractUpdates)>>> {
```

Comment:
Did you consider an alternate design where we separate peers, data and stats? Reason I ask is because this stream is doing a lot of different things:

- looping over peers
- reading from db (although you've abstracted this out to allow for testing I gues?)
- converting stream of responses into stream of state updates

Alternative components:

1. loop over peers, streaming responses
2. separate stream for fetching `StateUpdateStats`
3. zip (1) and (2), which turns response stream into state update stream

It does leave some issues where (1) cannot know that the stream has ended prematurely or not, so maybe its not good as is.. Could also be handled elsewhere.

Not super important atm, can leave as is. Just thought I'd get your thoughts.

Reply:
I like the idea, I'll raise an issue for a refactor because this PR is becoming to big imo.

> It does leave some issues where (1) cannot know that the stream has ended prematurely or not, so maybe its not good as is.. Could also be handled elsewhere.

I guess we can pass the head from the anchor to the stream "factory" method so that it'll error if ended prematurely. We can then pick up that error and restart the sync from that point.